ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committing/ProcessTransactionSubmission.scala

```diff
@@ -208,6 +212,63 @@ private[kvutils] case class ProcessTransactionSubmission(
           buildRejectionLogEntry(RejectionReason.Disputed("DuplicateKey: Contract Key not unique")))
     } yield r
 
+  /** Check that all parties mentioned in a transaction exist. */
+  private def validateParties: Commit[Unit] = {
+
+    def foldPartiesInTx[T](tx: GenTransaction.WithTxValue[NodeId, TContractId], z: T)(f: (T, String) => T): T =
+      foldPartiesInTxActors(tx, foldPartiesInTxVals(tx, z)(f))(f)
+
+    def foldPartiesInTxActors[T](tx: GenTransaction.WithTxValue[_, _], z: T)(f: (T, String) => T): T =
+      tx.fold(z) {
+        case (accum, (_, n)) => {
+          val parties = n match {
+            case c: NodeCreate[_, _] =>
+              (c.signatories union c.stakeholders)
+
+            case e: NodeExercises[_, _, _] =>
+              (e.actingParties union e.stakeholders)
+
+            case lk: Node.NodeLookupByKey[_, _] =>
+              lk.key.maintainers
+
+            case _ => Set[Ref.Party]()
+          }
+          parties.foldLeft(accum)(f)
+        }
+      }
+
+    def foldPartiesInTxVals[T](tx: GenTransaction.WithTxValue[_, TContractId], z: T)(f: (T, String) => T): T =
+      tx.foldValues(z) { (z, v) => foldPartiesInVal(v.value, z)(f) }
+
+    def foldPartiesInVal[T](v: Value[ContractId], z: T)(f: (T, String) => T): T =
```

Comment:
Is this stack-safe? Does it need to be?

Reply:
Not sure if there are protections uphill but, thinking locally, I can imagine this unbound stack to be a potential attack vector (even though non-malicious uses should be totally fine like this); there is also another one in `InputAndEffects.foldValue` so I'll turn them both `@tailrec`.